Partner Experience: update horizon config/install copy (#162)
stellar/stellar-docs: docs/run-platform-server/configuring.mdx

---
title: Configuring
sidebar_position: 30
---

import { Alert } from "@site/src/components/Alert";
import { CodeExample } from "@site/src/components/CodeExample";

Pre-requisites:

> - [Installation](./installing.mdx) is completed on host machine, it should have two executables, `stellar-horizon` and `stellar-core`. Per the installation instructions, `stellar-core` should be included in your host o/s PATH as these configurations will result in Horizon searching for `stellar-core` to execute.

Reviewer: ```suggestion
- [Installation](./installing.mdx) is completed on a host machine and results in two executables: `stellar-horizon` and `stellar-core`. Per the installation instructions, `stellar-core` should be included in your host OS' `$PATH`, as the configurations below result in Horizon searching for a `stellar-core` to execute.
```

- [Prepare the Database if first time](#prepare-the-database-first-time)

You are now ready to setup configuration parameters to perform three important, distinct roles:

- **serving requests** a regular web-based HTTP API,
- **ingesting ledgers** from core nodes of a Stellar network to keep its world-view up to date, and
- **transaction submission** for sending transactions to the Stellar network.

> You choose from one of two following deployment modes to perform the roles and proceed with configuration parameters to that means. A note on configuration parameters, the same parameter can be provided either as command line argument or an environment variable.

Reviewer: I find the "configuration parameters to that means" wording a bit confusing. Here's a suggested reword, plus moving the note to its own "Info" block:

```suggestion
To perform these roles, you can choose from one of two of the deployment modes that follow. Each has its own configuration parameters.

<Alert>

In all configuration examples, the same parameter can be provided either as command line argument or an environment variable. If both are provided, the command line takes precedence.

</Alert>

```

## Single Instance Deployment

> You run `stellar-horizon` in a single o/s process and it will perform all three roles at same time.

Reviewer: ```suggestion
In this simpler deployment variant, you run `stellar-horizon` in a single OS process and it will perform all three roles at same time.
```

Author: what's your opinion on using second person pronouns such as 'you'? briefly looking at tech/academic guides, they suggest rephrasing to 1st/3rd, the former looks widely used in docs, and assuming best to stay consistent for overall style?

Author: I'm not going to rock the boat on this, and follow existing precedence of referring to second person.

| flag               | envvar         | example                              |
| ------------------ | -------------- | ------------------------------------ |
| `--db-url`         | `DATABASE_URL` | postgres://localhost/horizon_testnet |

> | `--network`        | `NETWORK`      | pubnet                               |

Reviewer: Isn't it `--network-passphrase` and the full thing? Though this would be an awesome simplification.

Author: not quite, this is proposing a new flag `--network = [pubnet | testnet | other]` as a way to allow automating remainder of network specific flags and the network toml for captive core, I added a TODO with a bit more justification on using this new flag - https://github.com/stellar/stellar-docs/pull/162/files/6ae0bd3a50eccb4a7f8ea3ce99bb89d389321db5#diff-a2be184a631b87a98857de0dd0abe901235c4639179be7c62719b5ab4e6502ddR74

> | `--log-level`      | `LOG_LEVEL`    | info                                 |

Reviewer: I'm assuming this will be the default, so we can drop it to simplify the config?

> | `--tx-sub-enabled` | TX_SUB_ENABLED | true                                 |

Reviewer: this isn't a real flag, right?

Author: not yet, I'm proposing to add it as a TODO comment on https://github.com/stellar/stellar-docs/pull/162/files/6ae0bd3a50eccb4a7f8ea3ce99bb89d389321db5#diff-a2be184a631b87a98857de0dd0abe901235c4639179be7c62719b5ab4e6502ddR50

## Multiple Instance Deployment

> You run multiple instances of `stellar-horizon` each performs a subset of the roles.

Reviewer: ```suggestion
In this scalable deployment variant, you run multiple instances of `stellar-horizon` and each performs a subset of the roles.
```

### Ingestion Role Instance

> You need to allocate one instance to perform ongoing ingestion to capture network activity. We recommend limiting the amount of network activity to retain in the db on a sliding window of the last 30 days, we include `--history-retention-count 518400` which is number of ledgers in 30 days. **TODO - make --history-retention-count default to 518400, don't specify here anymore. make --captive-core-use-db=true by default**

Reviewer: ```suggestion
You need to allocate one instance to perform ongoing ingestion to capture network activity and store historical data. We recommend limiting the amount of historical data retained in the database on a sliding window of the last 30 days. This is controlled by the `--history-retention-count` parameter below, set to `518400` which is approximately the number of ledgers in 30 days. **TODO - make --history-retention-count default to 518400, don't specify here anymore. make --captive-core-use-db=true by default**
```

| flag | envvar | example |
| --- | --- | --- |
| `--db-url` | `DATABASE_URL` | postgres://localhost/horizon_testnet |
| `--history-retention-count` | `HISTORY_RETENTION_COUNT` | 518400 |
| `--network` | `NETWORK` | pubnet |
| `--log-level` | `LOG_LEVEL` | info |

### API Role Instance

> You can run none or multiple instances to serve read only API requests. **TODO - create --tx-sub-enabled flag, default=false. Should return 405 on POSTs to /tx if enabled. this allows skipping need to define --stellar-core-url on api servers since instance doesn't need to be aware of core.**

Reviewer: ```suggestion
You can run zero or more instances to serve read-only API requests. **TODO - create --tx-sub-enabled flag, default=false. Should return 405 on POSTs to /tx if enabled. this allows skipping need to define --stellar-core-url on api servers since instance doesn't need to be aware of core.**
```

| flag               | envvar           | example                              |
| ------------------ | ---------------- | ------------------------------------ |
| `--db-url`         | `DATABASE_URL`   | postgres://localhost/horizon_testnet |
| `--log-level`      | `LOG_LEVEL`      | info                                 |
| `--tx-sub-enabled` | `TX_SUB_ENABLED` | false                                |

### Transaction Submission Role Instance

> You can run none or multiple instances to serve transaction submission requests. If you run an instance with transaction submission enabled, yuur deployment is required to perform the ingestion role on the same database referenced here. Horizon transaction submission depends on ingestion to confirm tx submission status. **TODO - Confirm Horizon can use captive core to send tx sub requests and skip stellar-core-url?. If --tx-sub-enabled=true, horizon should check at startup that ingestion is happening on it's db or exit startup with error.**

Reviewer: ```suggestion
You can run zero or more instances to serve transaction submission requests. If you run an instance with transaction submission enabled, your deployment is **required** to perform the ingestion role on the same database referenced by this instance. Horizon's transaction submission depends on ingestion to confirm transaction submission status. **TODO - Confirm Horizon can use captive core to send tx sub requests and skip stellar-core-url?. If --tx-sub-enabled=true, horizon should check at startup that ingestion is happening on it's db or exit startup with error.**
```

| flag | envvar | example |
| --- | --- | --- |
| `--db-url` | `DATABASE_URL` | postgres://localhost/horizon_testnet |
| `--network` | `NETWORK` | pubnet |
| `--log-level` | `LOG_LEVEL` | info |
| `--stellar-core-url` | STELLAR_CORE_URL | http://localhost:11626 |
| `--tx-sub-enabled` | `TX_SUB_ENABLED` | true |

> ## Notes

Reviewer: imo this section should come before the example configs

- `--network`, optional, can be one of Stellar's public networks, 'pubnet', or 'devnet'. Triggers horizon to automatically set configurations for remaining horizon settings and generate the correct core toml/cfg settings. If you only need Horizon to connect to one of those public Stellar networks, this will take care of all related configuration. **TODO - create this new flag and the functionality it performs to configure all aspects including toml, archive urls, etc for network 'pubnet' or 'testnet'**

- If you want to connect Horizon to a different stellar network or override any of the defaults that horizon will make, the key configurations that can be set are: `--history-archive-urls`, `--captive-core-config-path`, `--network-pass-phrase`, `--captive-core-storage-path`, `--stellar-core-url`.

- `--db-url`, required, specifies the Horizon database. It's value follows this format: `dbname=<pg_dbname> user=<pg_user_name> password=<pg_user_pwd> host=<host_address>`

- `--log-level`, optional, can be one of 'info', 'error', 'debug'.

> - Horizon will create a sub-directory under the current working directory of the o/s process to store captive core runtime data files. Expect up to 10GB usage(**TODO - verify the max storage needs of core/buckets on disk?**). You can override this location with the optional `--captive-core-storage-path`, set to a directory on file system which captive core wiull store the runtime files.

Reviewer 2: ```suggestion
- Horizon will create a sub-directory under the current working directory of the o/s process to store captive core runtime data files. Expect up to 10GB usage(**TODO - verify the max storage needs of core/buckets on disk?**). You can override this location with the optional `--captive-core-storage-path`, set to a directory on file system which captive core will store the runtime files.
```

### Passing configurations to Horizon

> Depends on how Horizon was installed:

Reviewer: We should have complete sentences opening a paragraph,

```suggestion
Depending on how Horizon was installed, the way to configure it differs:
```

- bare-metal
  - non-package manager, we recommend using environment variables to pass configurations. There are many tools you can use to manage them, such as [direnv](http://direnv.net/) or [dotenv](https://github.com/bkeepers/dotenv).
  - [package manager](./installing.mdx#package-manager), the provided `stellar-horizon-cmd` wrapper will import a configuration from `/etc/default/stellar-horizon`, hence, if you want to set configurations, edit the file at `/etc/default/stellar-horizon`.
    <Alert>
      This script invokes Horizon with the `stellar` user, so make sure that
      permissions for the user are set up accordingly. The current working
      directory should be writable for this user and the user should be able to

>       execute the `horizon` and `stellar-core` binaries; etc.

Reviewer 2: Should this be "execute the `stellar-horizon`" instead since the name of the binary is stellar-horizon?

Reviewer 2: It looks like the binary name is "horizon" when downloaded as prebuilt binaries or built locally but it is "stellar-horizon" when installed via package manager. This distinction becomes relevant when users copy and paste example commands from our docs as it may not work correctly depending on the installation method. Previously the documentation had an additional step of renaming the binary using `sudo cp horizon /usr/local/bin/stellar-horizon` I'm uncertain about the reasoning behind the difference in names and the steps required to unify 
but mentioning it here as a potential TODO item (?).

    </Alert>
- containerized
  - non-Helm, pass all configuration parameters to the horizon docker image as [docker environment variables](https://docs.docker.com/engine/reference/commandline/run/#env).
  - Helm, pass all configuration parameters in the [Helm Install command](https://helm.sh/docs/helm/helm_install/)

## Prepare the Database first time

Before running the Horizon server first time, you must initialize the Horizon database. This database will be used for all of the information produced by Horizon, most notably historical information about transactions that have occurred on the Stellar network.

> To prepare a database for Horizon's use, first ensure it is blank. It's easiest to create a new database on your PostgreSQL server specifically for Horizon's use. We recommend creating a new user(role) in postgres dedicated for horizon's db and assigning that user(role) as the owner of the new horizon db. To illustrate this as example using `psql`, first login to databse server using the `psql` cli as a superuser, and then create the new user(role) and db for horizon:

Reviewer: ```suggestion
To prepare a database for Horizon's use, first ensure it is blank. It's easiest to create a new database on your PostgreSQL server specifically for Horizon's use. We recommend creating a new user(role) in postgres dedicated to Horizon's database and assigning that user(role) as the owner of this database. 

To illustrate an example using `psql`, first login to database server using the `psql` command-line tool as a superuser, and then create the new user(role) and database for Horizon:
```

```

> postgres=#

Reviewer 2: When I was setting up PostgreSQL locally I found the `createuser` and `createdb` utilities that come bundled with the installation of Postgres easy to use. 
1. Create a new user called 'horizon' with a password
`createuser -P horizon`  
2. Create a new database called 'horizon' with the user 'horizon' as the owner
 `createdb -O horizon horizon`

Reviewer 2: Thanks for providing the reasoning. It makes sense and I agree with you in that we should opt for the lowest common denominator.

postgres=# CREATE ROLE horizon WITH LOGIN;
CREATE ROLE
postgres=#
postgres=# CREATE DATABASE horizon OWNER horizon;
CREATE DATABASE
postgres=#
```

> Additionally set a password on your new `horizon` user using `ALTER USER`.

Reviewer: ```suggestion
Additionally, you can set a password on your new `horizon` user using `ALTER USER`.
```

> Given this example, you can now complete the `DATABASE_URL="dbname=horizon user=horizon password=<value_here> host=<address_of_pg_server>" configuration parameter.

Reviewer: code block isn't closed 

```suggestion
Given this example, you can now complete the `DATABASE_URL="dbname=horizon user=horizon password=<value_here> host=<address_of_pg_server>"` configuration parameter.
```

Author: nice catch, thanks.

Next, use the Horizon binary to install the schema onto the empty db from command line:

```
$ export DATABASE_URL="dbname=horizon user=horizon password=<value_here> host=<address_of_pg_server>"

> $ stellar-horizon db init --db-url

Reviewer: You don't need this since you `export`ed above:

```suggestion
$ stellar-horizon db init
```

---
title: Configuring
sidebar_position: 30
---

import { Alert } from "@site/src/components/Alert";
import { CodeExample } from "@site/src/components/CodeExample";

Pre-requisites:

- [Installation](./installing.mdx) is completed on a host machine and results in two executables: `stellar-horizon` and `stellar-core`. Per the installation instructions, `stellar-core` should be included in your host OS' `$PATH`, as the configurations below result in Horizon searching for a `stellar-core` to execute.
- [Initialize Database](#initialize-horizon-database)

You are now ready to setup configuration parameters to perform three important, distinct roles:

- **serving requests** via a regular web-based HTTP API, and

> - **ingesting ledgers** from core nodes of a Stellar network to keep its world-view up to date, and

Author: @mollykarcher , @urvisavla , @Shaptic , wondering about a possible fourth role..a 'current state' role, a horizon instance that runs ingestion w/captive core, but stores no history, just the current state. provides cohort of api endpoints that come with that data only such as order book, offers, trades(trade-agg?), etc, not sure if enough app uses cases fall into this.

the role would equate to lighter-weight deployment, likely not external postgres/db, maybe get by with transient sqlite db spun up internally on disk/mem, this may have been evaluated already.

Reviewer 2: >a horizon instance that runs ingestion w/captive core, but stores no history, just the current state.

Would this be same as running ingestion with no history retention? I just want to clarify my understanding of running ingestion with no history.

- **transaction submission** for sending transactions to the Stellar network.

To perform these roles, you can choose from one of two of the deployment modes that follow. Each has its own configuration parameters.

## Single Instance Deployment

You run `stellar-horizon` in a single o/s process and it will perform all three roles at same time.

| flag               | envvar         | example                              |
| ------------------ | -------------- | ------------------------------------ |
| `--db-url`         | `DATABASE_URL` | postgres://localhost/horizon_testnet |
| `--network`        | `NETWORK`      | pubnet                               |
| `--tx-sub-enabled` | TX_SUB_ENABLED | true                                 |

## Multiple Instance Deployment

You run multiple instances of `stellar-horizon` each performs a subset of the roles.

### Ingestion Role Instance

You need to allocate one instance to perform ongoing ingestion to capture network activity. This will default to limiting storage of ingested network activity in the database to a sliding window of the last 30 days.

> ** TODO - make --history-retention-count default to 518400 **

Author: yes, on second thought, will include recommendation for setting `--history-retention-count=518400` explicitly in docs, as the minor gain of not setting one parameter seems outweighed by the potential confusion that would be caused otherwise.

> ** TODO - make --captive-core-use-db=true by default **

Reviewer 2: This is addressed in https://github.com/stellar/go/issues/4856

If ingestion is planned to be done on a separate instance, add `INGEST=false` on this instance, otherwise don't include the parameter, Horizon will default to `INGEST=true` if absent.

**TODO - Confirm Horizon can use captive core to send tx sub requests and skip stellar-core-url?. **

**TODO - If --tx-sub-enabled=true, horizon should check at startup that ingestion is happening on it's db or exit startup with error.**

| flag               | envvar           | example                              |
| ------------------ | ---------------- | ------------------------------------ |
| `--db-url`         | `DATABASE_URL`   | postgres://localhost/horizon_testnet |
| `--network`        | `NETWORK`        | pubnet                               |
| `--tx-sub-enabled` | `TX_SUB_ENABLED` | true                                 |
| `--ingest`         | `INGEST`         | false                                |

## Notes

> - In all configuration examples, the same parameter can be provided either as command line argument or an environment variable. If both are provided, the command line takes precedence.

Reviewer 2: This clarification is great! I was wondering what happens if I specify both the command line and environment variable when setting up horizon first time locally. Also, I was curious if there are any advantages to choose one approach over the other and may be that's the only one we mention in our documentation (?).

Author: we suggest to use env variables across the board in 'passing config' section:

https://github.com/stellar/stellar-docs/pull/162/files#diff-a2be184a631b87a98857de0dd0abe901235c4639179be7c62719b5ab4e6502ddR100

I should add more context there to explain why, i.e. env variables are preferred all the time as they are more secure for obfuscating sensitive info that only the internal process can see, whereas command line leaves a trail that can be observed outside the process. It also depends on the install type, as far as what the operator/user interacts with, for bare metal, they define the env variables, but for containers, they use the container runtime mechanisms to pass env variables through such as docker/helm. 

@jacekn , in the case of Helm chart for Horizon, assuming install params would likely be passed as values file to `helm install ...` correct?

Reviewer 3: @sreuland yeah that right. We tried to make the chart default work out of the box with ability to tune them.
The chart is in early stages though so it can absolutely be improvement and defaults tuned if they are not right. They were based on what we run at SDF but those might not necessarily be the best ones for the community.

Author: ok, @urvisavla , can we capture this as dev ticket, 'deprecate command line input of horizon flags' , with an outcome of warnings printed by horizon if any cli present?

I'll do pass through here to rename ref's to any cli format of flags, to the environment variable format.

Author: except for `--help` , we should allow just that one cli flag

> - If you have configured your deployment to perform the ingestion role, then it is **strongly** recommended to review [Ingestion](./ingestion.mdx) first and [Filtering](./ingestion-filtering) second and factor that into configuration parameters to achieve best performance related to your application requirements before proceeding further.

Author: yes, needed to add a Notes bullet for `TX_SUB_ENABLED` in general, and reference to channel accounts usage. https://github.com/stellar/stellar-docs/pull/162/commits/8ba722ce02749aa17849e3bf3a5c1f342a0ebf3a

- `--network`, optional, can be one of Stellar's public networks, 'pubnet', or 'devnet'. Triggers horizon to automatically set configurations for remaining horizon settings and generate the correct core toml/cfg settings. If you only need Horizon to connect to one of those public Stellar networks, this will take care of all related configuration.

**TODO - create the new `--network` flag and the functionality it performs to configure all aspects of network connection, including toml, passphrase, archive urls, etc for networks 'pubnet' and 'testnet'. Refer to quickstart for example of same consolidated config parameter for setting network. **

> - If you want to connect Horizon to a different stellar network other than pubnet or testnet or override any of the defaults that `--network` usage will initiate, the key configuration flags that can be set are: `--history-archive-urls`, `--captive-core-config-path`, `--network-pass-phrase`, `--captive-core-storage-path`, `--stellar-core-url`.

Author: good point, swept that up - https://github.com/stellar/stellar-docs/pull/162/commits/a0f5d125ee71ece0e21d48dc0517dae3b34f3b13

> **TODO - create the new `NETWORK` env variable and the functionality it performs to configure all aspects of network connection, including toml, passphrase, archive urls, etc for networks 'pubnet' and 'testnet'. Refer to quickstart for example of same consolidated config parameter for setting network. **

Reviewer 2: I'm creating a dev ticket for this task and wanted to clarify a few points
1. If "NETWORK" is not specified, Horizon will still automatically set the values for CAPTIVE_CORE_CONFIG_PATH and CAPTIVE_CORE_STORAGE_PATH.

2. The CAPTIVE_CORE_STORAGE_PATH is not dependent on the network chosen. So the default storage path will remain the same regardless of the network.

Author: > If "NETWORK" is not specified, Horizon will still automatically set the values for CAPTIVE_CORE_CONFIG_PATH and CAPTIVE_CORE_STORAGE_PATH.

Do you see that behavior based on `NETWORK-PASSPHRASE`, I found an auto network config feature done in past based on that value, [go#3783](https://github.com/stellar/go/pull/3783), was not aware of this! I think the intent for its usage was more oriented to ease for dev, the design discussion struggled with prod usage, which we should learn from. 

@mollykarcher , I wonder how much this auto-config default behavior of `NETWORK-PASSPHRASE` is being relied upon in the field? It's not mentioned in the -help output, seems kind of quietly in there. Should we focus on shoring up the existing auto-config functionality of `NETWORK-PASSPHRASE` and referring to that usage here in docs or move forward with `NETWORK`? I kinda of prefer the latter as the former seems to conflate the meaning of what appears to be just a value-based setting, but need a check on this.

In either case for auto-config, I don't think Horizon should bother to configure default network or captive core settings if the key network settings are absent, as there is a valid configuration for the API role, where INGEST=false, TX_SUB_ENABLED=false and no network access is required, only the database is used. 

https://github.com/stellar/stellar-docs/pull/162/files/8ba722ce02749aa17849e3bf3a5c1f342a0ebf3a#diff-a2be184a631b87a98857de0dd0abe901235c4639179be7c62719b5ab4e6502ddR51

@urvisavla , this touches on need for new validations on config param settings that we should capture in the ticket also which we propose Horizon to detect/warn about, i.e. if INGEST=true or TX_SUB_ENABLED=true then network config settings must be present, either the key auto-config parameter or the collection of lower level params.
